homeassistant/helpers/condition.py

```diff
@@ -95,29 +98,30 @@ def async_from_config(
             )
         )
 
-    return cast(Callable[..., bool], factory(config, config_validation))
-
+    # Check for partials to properly determine if coroutine function
+    check_factory = factory
+    while isinstance(check_factory, ft.partial):
+        check_factory = check_factory.func
 
-from_config = _threaded_factory(async_from_config)
+    if asyncio.iscoroutinefunction(check_factory):
+        return cast(Callable[..., bool], await factory(hass, config, config_validation))
```

Comment:
Why are we making the factory async?

Reply:
Because the `async_device_from_config` factory needs to be async to call `async_get_integration`